### ISSUE TITLE
Issue with GeometricRV's "density"

### ISSUE THREAD

**Issue by peteroupc** [sympy/sympy#20031]
I have noticed the following issue with GeometricRV.

I can tell that GeometricRV begins the distribution at 1. (I am aware that there are several conventions for the geometric distribution in the literature; for example, another convention is whether a geometric r.v. is the number of failures before success or vice versa, and I am also aware that that distribution has a continuous counterpart.)  However, GeometricRV doesn't restrict its domain to integers (since GeometricRV is documented as _discrete_), and neither is the domain restricted to values that are 1 or greater.

As a result, the following unusual values occur with `density` for a GeometricRV:

```
In [38]: g=Geometric("G",p=S(1)/4)

In [39]: density(g)(0)
Out[39]: 1/3

In [40]: density(g)(-1)
Out[40]: 4/9

In [41]: density(g)(-2)
Out[41]: 
16
──
27

In [42]: density(g)(0.5)
Out[42]: 0.288675134594813

In [43]: density(g)(S(1)/2)
Out[43]: 
√3
──
6 

In [44]: density(g)(S(3)/2)
Out[44]: 
√3
──
8 

```

For all these, the value of the "density function" (more properly, mass function) should be 0, since the geometric value should be limited to integers greater than 1 if it's intended to be discrete have the mass function given in the documentation.

There may be other issues with GeometricRV, but this is the one I have noticed.

**Comment by Smit-create** [comment]
The reason for the incorrect result is this line:
https://github.com/sympy/sympy/blob/2573a46a40c20dcd5cd05cc114343f82a2f3869f/sympy/stats/drv_types.py#L114-L115

This can be fixed in the follwing way:
```diff
diff --git a/sympy/stats/drv_types.py b/sympy/stats/drv_types.py
index c0fb01946f..786f7d13ec 100644
--- a/sympy/stats/drv_types.py
+++ b/sympy/stats/drv_types.py
@@ -112,7 +112,8 @@ def check(p):
         _value_check((0 < p, p <= 1), "p must be between 0 and 1")

     def pdf(self, k):
-        return (1 - self.p)**(k - 1) * self.p
+        expr = (1 - self.p)**(k - 1) * self.p
+        return Piecewise((expr, self.set.as_relational(k)), (0, True))

```
I see this as a common issue with many distributions and requires similar changes for each of them, the results produced with the above diff:
```py
>>> from sympy.stats import *
>>> from sympy import S
>>> g = Geometric("G",p=S(1)/4)
>>> density(g)(0)
0
>>> density(g)(-1)
0
>>> density(g)(0.5)
0
>>> density(g)(S(1)/2)
0
>>> density(g)(S(3)/2)
0
```
I think this is easy to fix issue and can be fixed similarly for all the distributions

**Comment by sidhu1012** [comment]
Can I work on this issue?

**Comment by Smit-create** [comment]
Yes, sure. There is an open PR for Discrete RVs. If you wish you could start working on Continuous RVs

**Comment by sidhu1012** [comment]
>  If you wish you could start working on Continuous RVs
 The fix is same for continuous RVs or there's some differences to work on?

**Comment by sidhu1012** [comment]
> Yes, sure. There is an open PR for Discrete RVs. If you wish you could start working on Continuous RVs

Okay on it.

**Comment by sidhu1012** [comment]
@Smit-create pprint is not working in VS Code, can you pls help me with it.

**Comment by Maelstrom6** [comment]
> This can be fixed in the following way:
> 
> ```diff
> diff --git a/sympy/stats/drv_types.py b/sympy/stats/drv_types.py
> index c0fb01946f..786f7d13ec 100644
> --- a/sympy/stats/drv_types.py
> +++ b/sympy/stats/drv_types.py
> @@ -112,7 +112,8 @@ def check(p):
>          _value_check((0 < p, p <= 1), "p must be between 0 and 1")
> 
>      def pdf(self, k):
> -        return (1 - self.p)**(k - 1) * self.p
> +        expr = (1 - self.p)**(k - 1) * self.p
> +        return Piecewise((expr, self.set.as_relational(k)), (0, True))
> ```
> 
> I see this as a common issue with many distributions and requires similar changes for each of them

Should we do this for every distribution? Maybe it is better to have it implied as it already is. Making the pdfs more complicated than they need to be results in timeout problems for more complex tasks. See @sidhu1012's pull #20058 with two timeouts in [this test](https://travis-ci.org/github/sympy/sympy/jobs/725598182). All other failures are superficial and can be easily fixed since the master and his pull are both "correct" answers to the tests.

Many stats textbooks have the pdf implied to be 0 if x is not in the support. As a result, integrals over the real line are also implied to be integrals over the support. There are many other cases like this and I feel that it is simpler to keep it as it is. Maybe in the future, if `compound_rv.py` and others become far more efficient, then we could add these piecewise expressions.

**Comment by oscarbenjamin** [comment]
I think the problem is that it it really isn't a probability density function. It's a probability mass function (pmf).

If we want define probabilities for a discrete random variable in terms of integrals over a pdf then it should be a pdf made of Dirac delta functions.

**Comment by Maelstrom6** [comment]
You are correct, but what I said can be said for pmfs as well. Just replace "pdf" with "pmf" and "integrals" with "sums". 

Using the Dirac Delta also makes things more complicated than they need to be. In calculating expectations, working with a sum of evaluated pmfs is way simpler compared to an integral of a sum of Dirac Deltas.

**Comment by Smit-create** [comment]
I agree that `pdf` should be renamed as `pmf`.

**Comment by oscarbenjamin** [comment]
> what I said can be said for pmfs as well.

With a PMF there is no need to wrap it in a `Piecewise` as it is implied that the expression is only valid over a discrete state space. Also the issue here can be fixed since the `density` shouldn't be allowed for a PMF either.

**Comment by Smit-create** [comment]
> the `density` shouldn't be allowed for a PMF either.

`density()` is the only way the users could know the pmf/pdf of the distribution. It is like a common umbrella for pdf/pmf of all the Randoms variables.

**Comment by aleriv-umich** [comment]
Can I work on this issue?

**Comment by czgdp1807** [comment]
>  It is like a common umbrella for pdf/pmf of all the Randoms variables.

I believe changing this won't be a good idea as it will break a lot of code. Though it should be noted that by definition (see the quote from https://en.wikipedia.org/wiki/Probability_density_function) `density` is typically used for CRVs, and not FiniteRVs or DiscreteRVs. The question is if not density then what API should fetch the PMF for FiniteRVs and DiscreteRVs?

> density of a continuous random variable, is a function whose value at any given sample (or point) in the sample space (the set of possible values taken by the random variable) can be interpreted as providing a relative likelihood that the value of the random variable would equal that sample.

**Comment by Maelstrom6** [comment]
It seems that with the issues that @peteroupc is raising, a lot of these are inherently related to SymPy's bugs when integrating dirac deltas. I've run a debugger on most of his issues and the really problematic ones is due to said bug. The following issues are all because of this dirac delta problem and with other general problems with `Integrate`:
#17321
#19793
#19941
#20031
#20219

I have very little knowledge of the integrate side of SymPy. So, instead of fixing the problem with the dirac deltas, I have started a branch that changes `compute_density` in `rv.py`. [This is the comment](https://github.com/sympy/sympy/issues/20079#issuecomment-697559863) that I am trying to follow.

**Comment by czgdp1807** [comment]
> So, instead of fixing the problem with the dirac deltas, I have started a branch that changes `compute_density` in `rv.py`

Could you please post a link to that branch or a PR related to that branch?

**Comment by czgdp1807** [comment]
> The question is if not density then what API should fetch the PMF for FiniteRVs and DiscreteRVs?

We should keep `density` as an umbrella for both PDF and PMF because traditionally it is a part of `sympy.stats` API and changing it would create a lot of problems for the end users.

**Comment by czgdp1807** [comment]
Well, if we think carefully, then `density(g)` is nothing but a simple function with real domain. So, expecting it not to work outside the support of the RV is a bit vauge IMHO. In other words, its just a Garbage In Garbage Out situation. What should work correctly is queries like, `P(Eq(g, 0))`, `P(Eq(g, -1))`. 
```python
In [3]: g=Geometric("G",p=S(1)/4)
   ...: 

In [4]: density(g)(0)
Out[4]: 1/3

In [5]: P(Eq(g, 0))
Out[5]: 0

In [6]: P(Eq(g, 0.5))
Out[6]: 0

In [7]: P(Gt(g, 0.5))
Out[7]: 1
```

**Comment by Upabjojr** [comment]
There is already a `Density` class in `sympy.stats.rv`. `Density.doit()` returns the PDF.

One possibility the following:
1. in case of finite and discrete random symbols, `Density.doit()` should not evaluate or simply return a `Density` object.
2. add a method `__call__` to density, which should evaluate the density correctly (i.e. for finite RV check the support and multiply by Diract deltas)

I would suggest to avoid using PMF. In the end, PMF is simply a special case to represent densities with summations instead of integrals.

**Comment by Upabjojr** [comment]
Please do not edit every distribution... that would be a huge mess.

**Comment by czgdp1807** [comment]
> Please do not edit every distribution... that would be a huge mess.

I tried to find a fix without editing every distribution, opening a PR shortly.

**Comment by Maelstrom6** [comment]
> > So, instead of fixing the problem with the dirac deltas, I have started a branch that changes `compute_density` in `rv.py`
> 
> Could you please post a link to that branch or a PR related to that branch?

It's still a work in progress and has a lot further to go but [here it is](https://github.com/Maelstrom6/sympy/tree/stats_compute_independent_density).

**Comment by TanmayAgarwal123** [comment]
I would love to work on this issue, can you assign me
Regards,
Tanmay Agarwal

**Comment by KZV027** [comment]
May I suggest a potential modification to the **pdf** function to address the domain issue?

`def pdf(self, k):
    if not isinstance(k, int) or k < 1:  # Check for non-integer or negative input
        return 0  # Return 0 for invalid input
    else:
        return (1 - self.p)**(k - 1) * self.p  # Original calculation for valid input
`

**Comment by oscarbenjamin** [comment]
I think that there needs to be a separation between providing an expression for pdf/pmf that is implicitly valid only for the appropriate domain and separately providing something like a Piecewise that gives zero for expressions outside of the domain.

Expectation should be computed by summing or integrating over the appropriate domain and should not need to use a Piecewise expression to filter out values not in the domain.

**Comment by geodome** [comment]
Could this problem be solved through the assumption system such that if someone uses an invalid value that violates the assumption, Sympy would raise an exception instead of giving numbers?

**Comment by oscarbenjamin** [comment]
It can be resolved by returning a Piecewise. There should be a separation though between returning a Piecewise that might be hard to simplify vs returning an expresion that is implicitly only valid for the domain.

**Comment by geodome** [comment]
> > the `density` shouldn't be allowed for a PMF either.
> 
> `density()` is the only way the users could know the pmf/pdf of the distribution. It is like a common umbrella for pdf/pmf of all the Randoms variables.

could we have a mass() that essentially does the same thing as density but it would more logically coherent for the end-user to call it a PMF instead of PDF?

**Comment by geodome** [comment]
> > The question is if not density then what API should fetch the PMF for FiniteRVs and DiscreteRVs?
> 
> We should keep `density` as an umbrella for both PDF and PMF because traditionally it is a part of `sympy.stats` API and changing it would create a lot of problems for the end users.

What about adding mass() function that essentially does the same thing as density() so it would be logically coherent with the end-user's internal logic although the underlying mass() is actually calling density()? Also, extending the code by adding mass() won't break existing implementation that uses density() to invoke a PMF.

**Comment by geodome** [comment]
> It can be resolved by returning a Piecewise. There should be a separation though between returning a Piecewise that might be hard to simplify vs returning an expresion that is implicitly only valid for the domain.

Does this suffice as a valid patch?

`def pdf(self, k):
        return Piecewise((1 - self.p)**(k - 1) * self.p, And(k > 0, k <= floor(k)))`

**Comment by geodome** [comment]
> It can be resolved by returning a Piecewise. There should be a separation though between returning a Piecewise that might be hard to simplify vs returning an expresion that is implicitly only valid for the domain.

Does this suffice as a valid patch for the GeometricDistribution class? If yes, please assign this issue to me. Thanks.

```
    def pdf(self, k):
        return Piecewise(((1 - self.p)**(k - 1) * self.p, And(k > 0, k <= floor(k))), (0, True))
```

**Comment by oscarbenjamin** [comment]
While this can be "fixed" by using a Piecewise I think that it would be better not to make the pdf expression itself be a Piecewise. It is better to keep track separately of the expression for the pdf and the domain for which the expression is expected to be defined. Turning expressions into Piecewise makes it harder to simplify them later.

**Comment by geodome** [comment]
Accounting for the pdf in piecewise-form in the test script is not trivial.

**Comment by geodome** [comment]
>Turning expressions into Piecewise makes it harder to simplify them later.

I now appreciate this statement and understand its implication better.

**Comment by arhaque09** [comment]
I would like to work on this issue if its not been resolved yet!

**Comment by geodome** [comment]
> I would like to work on this issue if its not been resolved yet!

The problem is trickier than it looks. A simple rewrite of the pdf() will not pass the unit tests. I am studying the discrete random variable classes to discover a more elegant method to resolve this issue.

**Comment by geodome** [comment]
> I think that a better fix is like this:
> 
> ```diff
> diff --git a/sympy/stats/drv.py b/sympy/stats/drv.py
> index 13517e0f6d..5673ecd41e 100644
> --- a/sympy/stats/drv.py
> +++ b/sympy/stats/drv.py
> @@ -174,8 +174,14 @@ def expectation(self, expr, var, evaluate=True, **kwargs):
>              return Sum(expr * self.pdf(var),
>                           (var, self.set.inf, self.set.sup), **kwargs)
>  
> -    def __call__(self, *args):
> -        return self.pdf(*args)
> +    def __call__(self, arg):
> +        in_domain = self.set.contains(arg)
> +        if in_domain == False:
> +            return S.Zero
> +        elif in_domain == True:
> +            return self.pdf(arg)
> +        else:
> +            return Piecewise((self.pdf(arg), in_domain), (S.Zero, True))
>  
>  
>  class DiscreteDomain(RandomDomain):
> ```
> 
> When the density object is called with a value it is likely that `self.set.contains(arg)` will simplify to True or False and a Piecewise will only be made if it does not.
> 
> ```python
> In [9]: from sympy.stats import *
> 
> In [10]: g = Geometric("G",p=S(1)/4)
> 
> In [11]: density(g)
> Out[11]: GeometricDistribution(1/4)
> 
> In [12]: density(g)(1)
> Out[12]: 1/4
> 
> In [13]: density(g)(-1)
> Out[13]: 0
> 
> In [14]: density(g)(x)
> Out[14]: 
> ⎧   1 - x           
> ⎪4/3                
> ⎪────────  for x ∈ ℕ
> ⎨   4               
> ⎪                   
> ⎪   0      otherwise
> ⎩                   
> ```
> 
> The expression for pdf though will not be complicated by `Piecewise` which avoids making other calculations with pdf harder.
> 
> ```python
> In [15]: density(g).pdf(x)
> Out[15]: 
>    1 - x
> 4/3     
> ────────
>    4    
> 
> In [16]: density(g).set
> Out[16]: ℕ
> ```
> 
> The expression for the pdf should be understood as being valid only for elements of the set.
> 
> I don't know the classes in stats so well so it might be better to move the suggested code to a parent class or perhaps similar code should be added to other classes as well.

So you are saying the pdf() is actually wrapped by another function when Density calls it.

**Comment by oscarbenjamin** [comment]
It might be necessary to change some of the unit tests.

**Comment by arhaque09** [comment]
I will do that then!

**Comment by geodome** [comment]
> I will do that then!

Dude, I am working on this already since last week.

**Comment by geodome** [comment]
> It might be necessary to change some of the unit tests.

But rewriting unit tests is not always necessary the correct fix. There are concerns to ensure interoperability between the stats module and other Sympy modules. In this particular case, Piecewise gets in the way when attempting to unify the expressions even though it solves the immediate problem of giving the correct value.

**Comment by arhaque09** [comment]
> > I will do that then!
> 
> Dude, I am working on this already since last week. Can you work on other issues with no recent commit activities?

I apologize I will work on another project!

**Comment by oscarbenjamin** [comment]
Anyone is welcome to work on any issue. If someone is already working on something that is usually better to work together in the first instance rather than working independently though.

**Comment by Mike014** [comment]
Hi, I would like to work on this issue. I will review the code and submit a pull request soon. Let me know if there are any specific considerations. Thanks!

**Comment by Skyiesac** [comment]
PR :  [#28517](https://github.com/sympy/sympy/pull/28517)

**Comment by geodome** [comment]
good catch

**Comment by shuvro-git** [comment]
Just adding a small summary to help pinpoint the issue:

• The Geometric distribution is documented as a discrete RV taking values 1, 2, 3, …  
• But the current implementation does not restrict the domain to integers ≥ 1  
• As a result, density(g)(x) returns non-zero values for x ≤ 0 and for non-integers  
• This suggests that the mass function isn’t wrapped in a proper Piecewise with domain checks

Additionally, the second example shows an inconsistency in symbolic probability evaluation:

P(g1 - 1 = 0) = 3/4  
P(g2 - 1 = 0) = 3/4  
But P((g1 - 1) + (g2 - 1) = 0) incorrectly returns 0.

This may indicate an issue in how independence or integer domain constraints are handled in symbolic addition of discrete RVs.

Hope this helps narrow down the fix.
